_stbt/core.py

```diff
@@ -1979,11 +1998,7 @@ def on_new_sample(self, appsink):
                  "source-pipeline configuration." % sample.time)
 
         self.tell_user_thread(sample)
-        if self.lock.acquire(False):  # non-blocking
-            try:
-                self._sink_pipeline.push_sample(sample)
-            finally:
-                self.lock.release()
+        self._sink_pipeline.on_sample(sample)
```

Comment:
Before, `SinkPipeline.push_sample` would sometimes get called from the GLib main loop (i.e. the thread handling the GStreamer `new-sample` signal) and sometimes from the user thread that's running the user script (if using `stbt.frames()`, `stbt.wait_for_match()`, `stbt.wait_for_motion()`, or similar -- but *not* `stbt.get_frame()`).

N.B. `SinkPipeline.push_sample` is what draws on the frames. But previously if called from the GLib main loop there would be nothing to draw. Whereas now it will always do the drawing from the GLib main loop. Will that slow down the main loop from processing other events? Probably not. Just wondering out loud.

This way is certainly simpler & more consistent. I feel like this whole PR is making things simpler.

Reply:
> Before, SinkPipeline.push_sample would sometimes get called from the GLib main loop (i.e. the thread handling the GStreamer new-sample signal)

I think it would be called in the context of the "new-sample" signal from GStreamer.  This will be called from the GStreamer stream thread.  Bus messages are posted to the GLib main thread.  Signals are dispatched in-place (usually I think).

That's an interesting point about performance.  on-new-sample can't be dispatched again until we return from here.  Arguably the sink pipeline should have its own thread for this stuff.  I doubt that it would hurt too much though.  This won't be an issue if we ever get round to ditching the sink pipeline entirely. (one day...).